src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1156,6 +1156,50 @@ void MergeTreeData::clearOldTemporaryDirectories(ssize_t custom_directories_life
     }
 }
 
+MergeTreeData::DataPartsVector MergeTreeData::grabOldModifiedParts()
+{
+    DataPartsVector parts;
+
+    std::unique_lock lock(grab_old_modified_parts_mutex, std::defer_lock);
```

Comment:
We've discussed the problem with `BackgroundProcessingPool::addTask()` method (currently it is not possible to restrict execution of a task to only a signle worker from pool) with colleagues and are going to change this method.

In your case usage of atomic here is a much more proper way of serializing execution.